File: BinaryConverterJavaWebServer/src/java/ist/enesuysal/thesis/Receiver.java
package ist.enesuysal.thesis;

import ist.enesuysal.thesis.Helper.MyMethod;
import ist.enesuysal.thesis.Helper.MyField;
import ist.enesuysal.thesis.Helper.Helper;
import ist.enesuysal.thesis.Annotation.AvaliableMethod;
import ist.enesuysal.thesis.Annotation.Mandatory;
import static ist.enesuysal.thesis.CentralSerializer.convertToByteArray;
import ist.enesuysal.thesis.Tests.Test1;
import ist.enesuysal.thesis.Tests.Test2;
import ist.enesuysal.thesis.Tests.Test3;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.List;

public class Receiver {

    public MyMethod[] knownMethods = null;

    public Receiver() throws IllegalArgumentException, IllegalAccessException, InstantiationException, ClassNotFoundException {
        //Get ALL KNown AvaliableMethods
        List<Method> allAvaliableMethods = Helper.getMethodsAnnotatedWith(this.getClass(), AvaliableMethod.class);
        knownMethods = new MyMethod[allAvaliableMethods.size()];
        for (int i = 0; i < allAvaliableMethods.size(); i++) {
            knownMethods[i] = new MyMethod();
            Class myClass = allAvaliableMethods.get(i).getParameters()[0].getType();
            if (myClass.isPrimitive() || Helper.isWrapperType(myClass)) {
                knownMethods[i].myfields = convertToByteArray(Helper.GetFieldCode(allAvaliableMethods.get(i).getParameters()[0].getType().getTypeName()), new byte[0]);
            } else {

                Object objectInstance = myClass.newInstance();
                Field[] fields = myClass.getDeclaredFields();
                knownMethods[i].methodName = myClass.getName();
                byte[] fieldbytes = new byte[0];
                knownMethods[i].myfields = new byte[0];
                for (int j = 0; j < fields.length; j++) {
                    if (fields[j].isAnnotationPresent(Mandatory.class)) {
                        fieldbytes = CentralSerializer.serializePrimitive(fields[j].getType(), fields[j].getName(), true, fields[j].get(objectInstance), new byte[0]);
                    } else {
                        fieldbytes = CentralSerializer.serializePrimitive(fields[j].getType(), fields[j].getName(), false, fields[j].get(objectInstance), new byte[0]);
                    }
                    knownMethods[i].myfields = Helper.push(knownMethods[i].myfields, fieldbytes);
                }
            }
        }
    }

    public void createPrimitive(byte[] bytes) throws Exception {
        //Deserialize and Print
        //System.out.println("ss" + Arrays.toString(bytes));
        String type = (Helper.GetFieldType(bytes[0])); //FieldType
        int FieldNameLength = 0;
        byte[] FieldNameLengthByte = new byte[8];
        System.arraycopy(bytes, 1, FieldNameLengthByte, 0, FieldNameLengthByte.length);
        String fieldName = "";
        byte[] FieldNameByte = new byte[FieldNameLength];
        System.arraycopy(bytes, 9, FieldNameByte, 0, FieldNameByte.length);
        if (FieldNameLength > 0) {
            fieldName = CentralSerializer.convertToString(bytes);
        }
        //Field hasValue
        byte[] FieldValueByte = new byte[bytes.length - (17 + FieldNameLength)];
        System.arraycopy(bytes, 17 + FieldNameLength, FieldValueByte, 0, FieldValueByte.length);
        System.out.println("Value " + Helper.GetFieldValue(type, FieldValueByte));
        PrintObject(Helper.GetFieldValue(type, FieldValueByte));
    }

    public void createWrapper(byte[] bytes) throws Exception {
        //Deserialize and Print
        String type = (Helper.GetFieldType(bytes[1])); //FieldType
        int FieldNameLength = 0;
        byte[] FieldNameLengthByte = new byte[8];
        System.arraycopy(bytes, 2, FieldNameLengthByte, 0, FieldNameLengthByte.length);
        String fieldName = "";
        byte[] FieldNameByte = new byte[FieldNameLength];
        System.arraycopy(bytes, 10, FieldNameByte, 0, FieldNameByte.length);
        if (FieldNameLength > 0) {
            fieldName = CentralSerializer.convertToString(bytes);
        }
        //Field hasValue
        byte[] FieldValueByte = new byte[bytes.length - (10 + FieldNameLength)];
        System.arraycopy(bytes, 10 + FieldNameLength, FieldValueByte, 0, FieldValueByte.length);
        System.out.println("Wrapper Value " + Helper.GetFieldValue(type, FieldValueByte));
        PrintObject(Helper.GetFieldValue(type, FieldValueByte));
    }

    public void PrintObject(Object o) {
        StringBuilder result = new StringBuilder();
        String newLine = System.getProperty("line.separator");

        result.append(o.getClass().getName());
        result.append(" Object {");
        result.append(newLine);

        //determine fields declared in this class only (no fields of superclass)
        Field[] fields = o.getClass().getDeclaredFields();

        //print field names paired with their values
        for (Field field : fields) {
            result.append("  ");
            try {
                result.append(field.getName());
                result.append(": ");
                //requires access to private field:
                result.append(field.get(o));
            } catch (IllegalAccessException ex) {
                //System.out.println(ex);
            }
            result.append(newLine);
        }
        result.append("}");

        System.out.println(result.toString());
    }

    @AvaliableMethod
    public void MakeObjectA(int test) {

    }

    @AvaliableMethod
    public void MakeObjectA(byte test) {

    }

    @AvaliableMethod
    public void MakeObjectA(char test) {

    }

    @AvaliableMethod
    public void MakeObjectA(long test) {

    }

    @AvaliableMethod
    public void MakeObjectA(short test) {

    }

    @AvaliableMethod
    public void MakeObjectA(float test) {

    }

    @AvaliableMethod
    public void MakeObjectA(double test) {

    }

    @AvaliableMethod
    public void MakeObjectA(boolean test) {
    }

    @AvaliableMethod
    public void MakeObjectB(Byte test) {
    }
    @AvaliableMethod
    public void MakeObjectB(String test) {
    }

    @AvaliableMethod
    public void MakeObjectB(Integer test) {
    }

    @AvaliableMethod
    public void MakeObjectB(Boolean test) {
    }

    @AvaliableMethod
    public void MakeObjectB(Character test) {
    }

    @AvaliableMethod
    public void MakeObjectB(Short test) {
    }

    @AvaliableMethod
    public void MakeObjectB(Long test) {
    }

    @AvaliableMethod
    public void MakeObjectB(Double test) {
    }

    @AvaliableMethod
    public void MakeObjectB(Float test) {
    }

    @AvaliableMethod
    public void MakeObjectC(Test1 test) {
    }

}

File: CSharpWebServer/CSharpWebServer/ist.enesuysal.thesis/Receiver.cs
using CSharpWebServer.ist.enesuysal.thesis.Annotation;
using CSharpWebServer.ist.enesuysal.thesis.Tests;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;

namespace CSharpWebServer.ist.enesuysal.thesis
{
    public class Receiver
    {
        public List<MyMethod> knownMethods = null;
        public Receiver()
        {
            Type myType = (typeof(Receiver));
            // Get the public methods.
            MethodInfo[] myArrayMethodInfo = myType.GetMethods();
            knownMethods = new List<MyMethod>();
            for (int i = 0; i < myArrayMethodInfo.Length; i++)
            {

                object[] attributes = myArrayMethodInfo[i].GetCustomAttributes(typeof(AvaliableMethod), true);
                if (attributes.Length == 1)
                {

                    ParameterInfo o = myArrayMethodInfo[i].GetParameters()[0];
                    if (o.ParameterType.IsPrimitive || Helper.isWrapperType(o.ParameterType.ToString()))
                    {
                        knownMethods.Add(new MyMethod());
                        knownMethods[knownMethods.Count - 1].myfields = CentralSerializer.convertToByteArray(Helper.GetFieldCode(o.ParameterType.ToString()), new byte[0]);

                    }
                    else
                    {
                        Object objectInstance = Activator.CreateInstance(o.ParameterType);
                        FieldInfo[] fields = objectInstance.GetType().GetFields(); // Obtain all fields
                        knownMethods.Add(new MyMethod());
                        knownMethods[knownMethods.Count - 1].methodName = o.ParameterType.ToString();
                        byte[] fieldbytes = new byte[0];
                        knownMethods[knownMethods.Count - 1].myfields = new byte[0];
                        foreach (var field in fields) // Loop through fields
                        {
                            object[] Fattributes = field.GetCustomAttributes(typeof(Mandatory), true);
                            if (Fattributes.Length == 1)
                            {
                                fieldbytes = CentralSerializer.serializePrimitive(field.FieldType.FullName, field.Name, true, field.GetValue(objectInstance), new byte[0]);

                            }
                            else
                            {
                                fieldbytes = CentralSerializer.serializePrimitive(field.FieldType.FullName, field.Name, false, field.GetValue(objectInstance), new byte[0]);
                            }
                            knownMethods[knownMethods.Count - 1].myfields = Helper.push(knownMethods[knownMethods.Count - 1].myfields, fieldbytes);
                        }
                    }

                }
            }

        }
        public void createPrimitive(byte[] bytes)
        {
            //Deserialize and Print
            //System.out.println("ss" + Arrays.toString(bytes));
            String type = (Helper.GetFieldType(bytes[0])); //FieldType
            int FieldNameLength = 0;
            byte[] FieldNameLengthByte = new byte[8];
            Array.Copy(bytes, 1, FieldNameLengthByte, 0, FieldNameLengthByte.Length);
            String fieldName = "";
            byte[] FieldNameByte = new byte[FieldNameLength];
            Array.Copy(bytes, 9, FieldNameByte, 0, FieldNameByte.Length);
            if (FieldNameLength > 0)
            {
                fieldName = CentralSerializer.convertToString(bytes);
            }
            //Field hasValue
            byte[] FieldValueByte = new byte[bytes.Length - (17 + FieldNameLength)];
            Array.Copy(bytes, 17 + FieldNameLength, FieldValueByte, 0, FieldValueByte.Length);

            PrintObject(Helper.GetFieldValue(type, FieldValueByte), type);
        }

        public void createObject(String className, byte[] decodedBytes)
        {
            //Class c = Class.forName(className);
            //Object obj = c.newInstance();
            //Field[] methods = c.getFields();
            //for (Field f : methods) {
            //    f.set(obj, Helper.GetValue(decodedBytes, f.getName(), f.getType()));

            //}

            Type t = Type.GetType(className);
            object instance = Activator.CreateInstance(t);
            FieldInfo[] fields = t.GetFields(); // Obtain all fields
            foreach (var field in fields) // Loop through fields
            {
                string name = field.Name; // Get string name
                object temp = field.GetValue(instance); // Get value
                field.SetValue(instance, Helper.GetValue(decodedBytes, field.Name, field.FieldType.FullName)); //.set(obj, Helper.GetValue(decodedBytes, f.getName(), f.getType()));

            }
            PrintObject(instance);

        }

        public void PrintObject(Object o)
        {
            Type type = o.GetType(); // Get type pointer
            FieldInfo[] fields = type.GetFields(); // Obtain all fields
           
            foreach (var field in fields) // Loop through fields
            {
                string name = field.Name; // Get string name
                object temp = field.GetValue(o); // Get value
                Console.WriteLine("TYPE: " + field.FieldType.FullName);
                Console.WriteLine("Name: " + name);
                Console.WriteLine("Value: " + temp);
            }
        }

        public void PrintObject(Object o, string type)
        {
            Console.WriteLine("TYPE: " + type);
            Console.WriteLine("Value: " + o);
        }
        [AvaliableMethod]
        public void MakeObjectA(int test)
        {

        }

        [AvaliableMethod]
        public void MakeObjectA(byte test)
        {

        }

        [AvaliableMethod]
        public void MakeObjectA(Char test)
        {

        }

        [AvaliableMethod]
        public void MakeObjectA(long test)
        {

        }

        [AvaliableMethod]
        public void MakeObjectA(short test)
        {

        }

        [AvaliableMethod]
        public void MakeObjectA(float test)
        {

        }

        [AvaliableMethod]
        public void MakeObjectA(double test)
        {

        }

        [AvaliableMethod]
        public void MakeObjectA(bool test)
        {
        }


        [AvaliableMethod]
        public void MakeObjectB(string test)
        {
        }


        [AvaliableMethod]
        public void MakeObjectB(short test)
        {
        }


        [AvaliableMethod]
        public void MakeObjectB(double test)
        {
        }

        [AvaliableMethod]
        public void MakeObjectB(float test)
        {
        }

        [AvaliableMethod]
        public void MakeObjectC(Test1 test)
        {
        }

        [AvaliableMethod]
        public void MakeObjectC(Test2 test)
        {
        }

        [AvaliableMethod]
        public void MakeObjectC(Test3 test)
        {
        }


    }
}
